Kernel/Syscalls/fallocate.cpp

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2022, Leon Albrecht <leon.a@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Checked.h>
+#include <Kernel/FileSystem/Inode.h>
+#include <Kernel/FileSystem/InodeFile.h>
+#include <Kernel/FileSystem/OpenFileDescription.h>
+#include <Kernel/Process.h>
+
+namespace Kernel {
+
+ErrorOr<FlatPtr> Process::sys$posix_fallocate(int fd, Userspace<off_t const*> userspace_offset, Userspace<off_t const*> userspace_length)
+{
+    VERIFY_NO_PROCESS_BIG_LOCK(this);
+    TRY(require_promise(Pledge::stdio));
+
+    auto offset = TRY(copy_typed_from_user(userspace_offset));
+    if (offset < 0)
+        return EINVAL;
+    auto length = TRY(copy_typed_from_user(userspace_length));
+    if (length <= 0)
+        return EINVAL;
+
+    Checked<size_t> checked_size { length };
+    checked_size += offset;
+    // FIXME: Return EFBIG if offset+length > FileSizeMax
+    if (checked_size.has_overflow())
+        return EFBIG;
+
+    auto description = TRY(open_file_description(fd));
+    if (!description->is_writable())
+        return EBADF;
+
+    if (description->is_fifo())
+        return ESPIPE;
+
+    if (!S_ISREG(TRY(description->file().stat()).st_mode))
+        return ENODEV;
+
+    auto& file = static_cast<InodeFile&>(description->file());
```

Comment:
Do we maybe want to do `VERIFY(description->file().is_inode())` before this regardless? I don't think that this will ever be hit anyways given the file types, but it seems like a very inexpensive check that would save us from tracking down some really hard to debug issues in case we accidentally do a wrong cast.